pkg/batchscheduler/volcano/volcano_scheduler.go

```diff
@@ -0,0 +1,184 @@
+	volcanoClient   volcanoclient.Interface
+}
+
+func GetPluginName() string {
+	return "volcano"
+}
+
+func (v *VolcanoBatchScheduler) Name() string {
+	return GetPluginName()
+}
+
+func (v *VolcanoBatchScheduler) ShouldSchedule(app *v1beta1.SparkApplication) bool {
+
+	checkScheduler := func(scheduler *string) bool {
+		return scheduler != nil && *scheduler == v.Name()
+	}
+
+	if app.Spec.Mode == v1beta1.ClientMode {
+		return checkScheduler(app.Spec.Executor.SchedulerName)
+	}
+	if app.Spec.Mode == v1beta1.ClusterMode {
+		return checkScheduler(app.Spec.Executor.SchedulerName) && checkScheduler(app.Spec.Driver.SchedulerName)
+	}
+
+	glog.Warningf("Unsupported Spark application mode %s, abandon schedule via volcano.", app.Spec.Mode)
+	return false
+}
+
+func (v *VolcanoBatchScheduler) OnSparkApplicationSubmitted(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	newApp := app.DeepCopy()
+	if newApp.Spec.Executor.Annotations == nil {
+		newApp.Spec.Executor.Annotations = make(map[string]string)
+	}
+
+	if newApp.Spec.Driver.Annotations == nil {
+		newApp.Spec.Driver.Annotations = make(map[string]string)
+	}
+
+	if newApp.Spec.Mode == v1beta1.ClientMode {
+		return v.syncPodGroupInClientMode(newApp)
+	} else if newApp.Spec.Mode == v1beta1.ClusterMode {
+		return v.syncPodGroupInClusterMode(newApp)
+	}
+	return newApp, nil
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroupInClientMode(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	//We only care about the executor pods in client mode
+	newApp := app.DeepCopy()
+	if _, ok := newApp.Spec.Executor.Annotations[v1alpha2.GroupNameAnnotationKey]; !ok {
+		if err := v.syncPodGroup(newApp, *newApp.Spec.Executor.Instances); err == nil {
+			newApp.Spec.Executor.Annotations[v1alpha2.GroupNameAnnotationKey] = v.getAppPodGroupName(newApp)
+		} else {
+			return nil, err
+		}
+	}
+	return newApp, nil
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroupInClusterMode(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	//We need both mark Driver and Executor when submitting
+	//NOTE: In cluster mode, the initial size of PodGroup is set to 1 in order to **only** schedule driver pod,
+	//and once the driver pod get scheduled, we will update the PodGroup size to reflect the actual size.
+	if _, ok := app.Spec.Driver.Annotations[v1alpha2.GroupNameAnnotationKey]; !ok {
+		if err := v.syncPodGroup(app, 1); err == nil {
+			app.Spec.Executor.Annotations[v1alpha2.GroupNameAnnotationKey] = v.getAppPodGroupName(app)
+			app.Spec.Driver.Annotations[v1alpha2.GroupNameAnnotationKey] = v.getAppPodGroupName(app)
+		} else {
+			return nil, err
+		}
+	}
+	return app, nil
+}
+
+func (v *VolcanoBatchScheduler) OnSparkDriverPodScheduled(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	// Update PodGroup MinMember if required
+	if app.Spec.Mode == v1beta1.ClusterMode {
+		return app, v.syncPodGroup(app, 1+*(app.Spec.Executor.Instances))
```

Comment:
why set this? btw, where do we set `minResources`?

Reply:
hmm, catch up with your conversion now.

Reply:
ok, this is what I infer from your conversation, 
1. For client Mode: We will schedule jobs when  `pod resource > minMember and free resource > resource(executor * executerInstance ) `
2. For driver Mode: We will schedule jobs when  `pod resource > 1 (for driver)  and free resource >resource(executor * executerInstance + driver)`

@k82cn  Do I get the correct conclusion?